experimental/torch_xla2/torch_xla2/tensor.py

```diff
@@ -281,3 +239,155 @@ def debug_accuracy(func, args, kwargs, current_output):
       pdb.set_trace()
 
   return True
+
+
+class XLAFunctionMode(torch.overrides.TorchFunctionMode):
+  """Context manager that dispatches torch function calls to JAX."""
+
+  def __init__(self, env):
+     self.env = env
+
+  def __torch_function__(self,
+                         func,
+                         types,
+                         args=(),
+                         kwargs=None) -> torch.Tensor:
+    try:
+      return self.env.dispatch(func, types, args, kwargs)
+    except OperatorNotFound:
+      return func(*args, **(kwargs or {}))
+
+
+class XLADispatchMode(torch_dispatch.TorchDispatchMode):
+
+  def __init__(self, env):
+    self.env = env
+
+  def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+    if isinstance(func, torch._ops.OpOverloadPacket):
+      with self:
+        return func(*args, **kwargs)
+    if func.namespace != 'aten':
+      return func(*args, **kwargs)
+    return self.env.dispatch(func, types, args, kwargs)
+
+def _name_of_func(func):
+  if hasattr(func, 'name'):
+    return func.name()
+  return func.__name__
+
+
+class Environment:
+    """This class holds a set of configurations and "globals" needed
+
+    for executing torch program using jax.
+    Things included so far:
+
+    op registry
+    PRNGKey
+    Configs
+
+    Also helper functions to manipulate those.
+    """
+
+    _prng_key: jax.random.PRNGKey
```

Comment:
I have a general design question about how to handle randomness. I had a half-finished addition for a few random ops. In there, I ignored the Environment's PRNG key entirely.

The most important thing for me personally is to preserve as much of `torch`'s semantics as possible. So for a function like `randn`, we should support the `generator` argument as the way for a user to pass in a key. We can call `Generator.seed` and then convert that to a JAX key.



Since we would have to support `generator` anyway in that case, it occurred to me that we don't actually need an extra way to store the global/default PRNG state -- `torch` already gives us one through `torch.random`, and it's already in use in much existing `torch` code (e.g. everyone is already calling `torch.manual_seed`). Do we also want to add an additional random state on top of that?